---
title: "bulk_RNAseq_QC"
author: "ENPRC Gencore"
date:  "`r Sys.Date()`"
output: 
  rmdformats::robobook:
    fig_width: 9
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE,
                      message = FALSE, 
                      echo=FALSE,
                      fig.width = 6,
                      fig.height = 4,
                      dev='png')
library(tidyverse)    ## General logic and processing
library(yaml)         ## Parse config
library(DESeq2)       ## Makes us have a job
library(tools)
#library(circlize)    ## For colorRamp2 in heatmaps
library(openxlsx)     ## Writing count table outputs
library(here)         ## For consistent directory structures
##library(plotly)     ## If you want interactive PCA
library(ggrepel)      ## for PCA plot repel text
# library(plyr)
library(fgsea)
library(msigdbr)      ## Loads pathway genesets from MsigDB
library(patchwork)


##remotes::install_github('yerkes-gencore/gencore-bulk',force = TRUE)
library(gencoreBulk)  ## See the git repo for instructions on using this

source("/yerkes-cifs/runs/home/gktharp/gitrepos/Coding_Club/BulkRNASeq/gktharp/GKTDESeq2Extentions.R")


## sometimes they conflict with other loaded packages
counts <- DESeq2::counts
here <- here::here
```



```{r reset_analysis_obj, eval=TRUE, include=FALSE}
# This is moved out of the setup block to avoid accidentally overwriting your
# analysis object when restarting a session
analysis = list()
```

# Load config

```{r load_config, include=FALSE}
# Store config with results object for record keeping
analysis$qc_config <- yaml.load_file(here("config/QC_config.yml"))
```

# Load samplesheet

```{r sample_metadata_setup}
## See the helper_script generate_experimental_design_sheet.R to help generate a table,
## or modify the template in the config folder and save it as 'exp_design'
# analysis$sampleTable <- read.table(here("config/exp_design.txt"),
#                                    header=T, 
#                                    stringsAsFactors = TRUE)

analysis$sampleTable <- tibble(
  FileID=dir(analysis$qc_config$alignmentDir)) %>%
  mutate(SampleID = str_sub(FileID, start = 13),
         SubjectID = str_extract(SampleID,"[MN][^-]{3}"),
         Timepoint = str_extract(SampleID,"Day[:digit:]{1,2}"),
         Group = factor(str_extract(SampleID,"ART.*"), levels = c(unique(str_extract(SampleID,"ART.*")))),
         Venetoclax = factor(if_else(Group == "ART","None","Venetoclax"),levels = c("None","Venetoclax")))

## Possible types of processing to arrange or factorize terms
#analysis$sampleTable <- analysis$sampleTable %>% dplyr::arrange(across(all_of(analysis$qc_config$sampleGrouping)))
# analysis$sampleTable <- analysis$sampleTable %>%
#   mutate(across(.), as.factor)

sample_metadata_table <- knitr::kable((analysis$sampleTable), position='center', table.attr="style='width:100%;'", format='html')# %>%
    #kable_styling(full_width = TRUE)
# Store the order you'd like samples to appear in figure
# Order is determined by sampleGrouping parameters in config
# But you could also manually specify it here if there aren't easy rules
# sample_order <- fct_inorder(as.character(analysis$sampleTable$SampleID))

## for compatibility with filepaths
analysis$sampleTable$FileID <- as.character(analysis$sampleTable$FileID)

sample_metadata_table
```

# Load counts tables

```{r read_counts_tables, include=FALSE}
STARreadSuffix <- "_ReadsPerGene.out.tab"
readfiles <- sapply(
      analysis$sampleTable$FileID,
      function(sid) {
       paste0(dir(analysis$qc_config$alignmentDir,
                  pattern = sid, full.names = TRUE),
              "/", sid, STARreadSuffix)})
          
outs <- parseReadPerGeneFiles(readfiles, analysis$qc_config$libraryType)

mapBins <- outs$map_bins
raw_counts <- outs$read_counts

# order counts as they are in the sampletable
raw_counts <- raw_counts[,analysis$sampleTable$FileID]
```

# Load GTF

```{r gene_names, include=TRUE}
geneInfoTab <- read_tsv(file.path(analysis$qc_config$referenceDir,
                                  "geneInfo.tab"),
                        skip = 1,
                        col_names = c("gene_id","gene_symbol","gene_type"))

## Remove samples if there are fewer total fragment counts than features
## could catch a technical error if data isn't loading correctly
## (but you should see this in the mapping plot)
lowcounts <- colSums(raw_counts) < nrow(geneInfoTab)
stopifnot(!sum(lowcounts)>0)

gene_names <- plyr::mapvalues(rownames(raw_counts), 
                        from=geneInfoTab$gene_id,
                        to=geneInfoTab$gene_symbol)

## save this in case you want to write it out
raw_counts_ensembl <- raw_counts
rownames(raw_counts) <- make.names(gene_names, unique=TRUE)
```

# Create DESeq object

```{r create_dds, include=FALSE}
## safety check to make sure metadata and samples are properly associated
stopifnot(colnames(raw_counts) == analysis$sampleTable$FileID)

analysis$dds <- DESeqDataSetFromMatrix(raw_counts,
                                       analysis$sampleTable,
                                       design = as.formula( ~ Venetoclax))
```

# Mapping bins plots
```{r mapping_plot, fig.width=8, fig.height=6}
analysis$plots$mapping_bins_plot <- mappingBinsPlot(mapBins)
analysis$plots$mapping_bins_plot
```

<!-- # Remove samples -->

<!-- ```{r remove_samples} -->
<!-- ## Removing samples specified in the config.  -->
<!-- ## This occurs after the mapBin plot just to have record of the  -->
<!-- ## removed sample and possibly why it was removed,  -->
<!-- ## but it could be adjusted to be upstream. -->

<!-- if (!is.null(unlist(analysis$qc_config$dropSamples))){ -->
<!--   mapBins <- mapBins[ -->
<!--     ,!colnames(mapBins) %in% analysis$qc_config$dropSamples] -->
<!--   analysis$dds <- analysis$dds[ -->
<!--     ,!(colnames(analysis$dds) %in% analysis$qc_config$dropSamples)] -->
<!--   # sample_order <- sample_order[ -->
<!--   #   !(sample_order %in% as.factor(analysis$qc_config$dropSamples))] -->
<!--   out <- paste0("Sample(s)\n", -->
<!--                 paste0(analysis$qc_config$dropSamples, sep=", ", collapse = ''), -->
<!--                 "\nbeing removed from downstream analysis per user specification.")  -->
<!--     cat(out) -->
<!-- }  -->
<!-- ``` -->

# VST and RLD

```{r vst_and_ddsDrop}
analysis$dds <- DESeq(analysis$dds, parallel = TRUE)
assays(analysis$dds)$vst <- varianceStabilizingTransformation(analysis$dds,
                                                         blind = FALSE,
                                                         fitType = "parametric")

assays(analysis$dds)$rld <- rlog(analysis$dds,
                                 blind = FALSE,
                                 fitType = "parametric")
```

# RLE plots

Library size normalization of read counts with hypothesis that most genes are 
not differentially expressed

```{r rle, fig.width = 6, fig.height=11}
rle_plots <- checkRLE(analysis$dds)
analysis$plots$RLE_raw <- rle_plots$RLE_raw
analysis$plots$RLE_norm <- rle_plots$RLE_norm
analysis$plots$RLE_raw
analysis$plots$RLE_norm
```

# PCA

Exploring projections of expression profiles onto low dimensional space using a 
limited number of highly variable features. This will give us an estimate of 
similarity between samples, but will not account for all genes.

```{r pca, fig.width=12, fig.height=8, warning=FALSE}
analysis$plots$pca <- plotPCAFromConfig(analysis)
print(analysis$plots$pca)
```

# Results

## Venetoclax

```{r}

# out of 23298 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0 (up)       : 1279, 5.5%
# LFC < 0 (down)     : 638, 2.7%
# outliers [1]       : 15, 0.064%
# low counts [2]     : 8415, 36%
# (mean count < 2)
# [1] see 'cooksCutoff' argument of ?results
# [2] see 'independentFiltering' argument of ?results

resVenetoclax <- results(analysis$dds,
                         name = "Venetoclax_Venetoclax_vs_None",
                         alpha = 0.05,
                         filter = maxMinFilter(analysis$dds,
                                               intgroup = "Venetoclax",
                                               comp = c("None","Venetoclax"))
                         )
summary(resVenetoclax)
```

```{r}
as.data.frame(resVenetoclax) %>%
  filter(!is.na(padj)) %>%
  mutate(DE = case_when(
    padj >= 0.05 ~ "not DE",
    log2FoldChange > 0 ~ "Up",
    log2FoldChange < 0 ~ "Down"
  ) %>% factor(levels = c("Up","Down","not DE"))) %>%
  group_by(DE) %>%
  summarise(n = n())
  
```


## Group

```{r}

analysis$ddsGroup <- DESeqDataSetFromMatrix(counts(analysis$dds),
                                            select(as.data.frame(colData(analysis$dds)),
                                                   -c(sizeFactor,replaceable)),
                                            design = ~ Group)
analysis$ddsGroup <- DESeq(analysis$ddsGroup, parallel = TRUE)

```

```{r}
resAVvsA <- results(analysis$ddsGroup,
                name = "Group_ARTposVenetoclax_vs_ART",
                alpha = 0.05,
                filter = maxMinFilter(analysis$ddsGroup,
                                      intgroup = "Group",
                                      comp = c("ART","ARTposVenetoclax")))

resAVCvsA <- results(analysis$ddsGroup,
                name = "Group_ARTposVenetoclaxposantiposCD8_vs_ART",
                alpha = 0.05,
                filter = maxMinFilter(analysis$ddsGroup,
                                      intgroup = "Group",
                                      comp = c("ART","ARTposVenetoclaxposantiposCD8")))

resAVCvsAV <- results(analysis$ddsGroup,
                contrast = c("Group","ARTposVenetoclaxposantiposCD8","ARTposVenetoclax"),
                alpha = 0.05,
                filter = maxMinFilter(analysis$ddsGroup,
                                      intgroup = "Group",
                                      comp = c("ARTposVenetoclaxposantiposCD8",
                                               "ARTposVenetoclax")))
```

```{r}

summary(resAVvsA)
as.data.frame(resAVvsA) %>%
  filter(!is.na(padj)) %>%
  mutate(DE = case_when(
    padj >= 0.05 ~ "not DE",
    log2FoldChange > 0 ~ "Up",
    log2FoldChange < 0 ~ "Down"
  ) %>% factor(levels = c("Up","Down","not DE"))) %>%
  group_by(DE) %>%
  summarise(n = n())
  
summary(resAVCvsA)
as.data.frame(resAVCvsA) %>%
  filter(!is.na(padj)) %>%
  mutate(DE = case_when(
    padj >= 0.05 ~ "not DE",
    log2FoldChange > 0 ~ "Up",
    log2FoldChange < 0 ~ "Down"
  ) %>% factor(levels = c("Up","Down","not DE"))) %>%
  group_by(DE) %>%
  summarise(n = n())
  
summary(resAVCvsAV)
as.data.frame(resAVCvsAV) %>%
  filter(!is.na(padj)) %>%
  mutate(DE = case_when(
    padj >= 0.05 ~ "not DE",
    log2FoldChange > 0 ~ "Up",
    log2FoldChange < 0 ~ "Down"
  ) %>% factor(levels = c("Up","Down","not DE"))) %>%
  group_by(DE) %>%
  summarise(n = n())
  

```


```{r fig.height=8, fig.width=10}
volD25VAvsA <- as.data.frame(resVenetoclax) %>%
  filter(!is.na(padj)) %>%
  mutate(DE = case_when(
    padj >= 0.05 ~ "12951",
    log2FoldChange > 0 ~ "1279",
    log2FoldChange < 0 ~ "638"
  ) %>% factor(levels = c("1279","638","12951"))) %>%
  mutate(Significant=padj<0.05) %>%
  mutate(Significant = factor(Significant,levels = c(TRUE,FALSE))) %>%
  ggplot(aes(x=log2FoldChange,y=-log10(pvalue),color = DE)) +
  geom_hline(aes(yintercept = -log10(pForAlpha(resVenetoclax)),
             linetype = "padj = 0.05"),color = "green") +
  geom_point(size = 1) +
  ylim(0,22.2) +
  scale_color_manual(values = c("1279" = "red","638" = "blue","12951" = "grey"),breaks = c("1279","638")) +
  scale_linetype_manual(name = "Cutoff",
                        values = c(2),
                        guide = guide_legend(override.aes = list(color = c("green")))) +
  guides(
    color = guide_legend(order = 1),
    linetype = guide_legend(order = 2)
  ) +
  theme_bw() +
  ggtitle("VA/A D25")



volD25AVvsA <- as.data.frame(resAVvsA) %>%
  filter(!is.na(padj)) %>%
  mutate(DE = case_when(
    padj >= 0.05 ~ "12767",
    log2FoldChange > 0 ~ "1156",
    log2FoldChange < 0 ~ "405"
  ) %>% factor(levels = c("1156","405","12767"))) %>%
  mutate(Significant=padj<0.05) %>%
  mutate(Significant = factor(Significant,levels = c(TRUE,FALSE))) %>%
  ggplot(aes(x=log2FoldChange,y=-log10(pvalue),color = DE)) +
  geom_hline(aes(yintercept = -log10(pForAlpha(resAVvsA)),
             linetype = "padj = 0.05"),color = "green") +
  geom_point(size = 1) +
  ylim(0,22.2) +
  scale_color_manual(values = c("1156" = "red","405" = "blue","12767" = "grey"),breaks = c("1156","405")) +
  scale_linetype_manual(name = "Cutoff",
                        values = c(2),
                        guide = guide_legend(override.aes = list(color = c("green")))) +
  guides(
    color = guide_legend(order = 1),
    linetype = guide_legend(order = 2)
  ) +
  theme_bw() +
  ggtitle("AV/A D25")

volD25AVCvsA <- as.data.frame(resAVCvsA) %>%
  filter(!is.na(padj)) %>%
  mutate(DE = case_when(
    padj >= 0.05 ~ "13877",
    log2FoldChange > 0 ~ "217",
    log2FoldChange < 0 ~ "161"
  ) %>% factor(levels = c("217","161","13877"))) %>%
  mutate(Significant=padj<0.05) %>%
  mutate(Significant = factor(Significant,levels = c(TRUE,FALSE))) %>%
  ggplot(aes(x=log2FoldChange,y=-log10(pvalue),color = DE)) +
  geom_hline(aes(yintercept = -log10(pForAlpha(resAVCvsA)),
             linetype = "padj = 0.05"),color = "green") +
  geom_point(size = 1) +
  ylim(0,22.2) +
  scale_color_manual(values = c("217" = "red","161" = "blue","13877" = "grey"),breaks = c("217","161")) +
  scale_linetype_manual(name = "Cutoff",
                        values = c(2),
                        guide = guide_legend(override.aes = list(color = c("green")))) +
  guides(
    color = guide_legend(order = 1),
    linetype = guide_legend(order = 2)
  ) +
  theme_bw() +
  ggtitle("AVC/A D25")

volD25AVCvsAV <- as.data.frame(resAVCvsAV) %>%
  filter(!is.na(padj)) %>%
  mutate(DE = case_when(
    padj >= 0.05 ~ "16048",
    log2FoldChange > 0 ~ "6",
    log2FoldChange < 0 ~ "4"
  ) %>% factor(levels = c("6","4","16048"))) %>%
  mutate(Significant=padj<0.05) %>%
  mutate(Significant = factor(Significant,levels = c(TRUE,FALSE))) %>%
  ggplot(aes(x=log2FoldChange,y=-log10(pvalue),color = DE)) +
  geom_hline(aes(yintercept = -log10(pForAlpha(resAVCvsAV)),
             linetype = "padj = 0.05"),color = "green") +
  geom_point(size = 1) +
  ylim(0,22.2) +
  scale_color_manual(values = c("6" = "red","4" = "blue","16048" = "grey"),breaks = c("6","4")) +
  scale_linetype_manual(name = "Cutoff",
                        values = c(2),
                        guide = guide_legend(override.aes = list(color = c("green")))) +
  guides(
    color = guide_legend(order = 1),
    linetype = guide_legend(order = 2)
  ) +
  theme_bw() +
  ggtitle("AVC/AV D25")

volD25VAvsA + volD25AVvsA + volD25AVCvsA + volD25AVCvsAV + patchwork::plot_layout(ncol=2)

```

### Modified volcano for manuscript

```{r fig.height=8, fig.width=10}

#volD25VAvsA <- 
as.data.frame(resVenetoclax) %>%
  filter(!is.na(padj)) %>%
  mutate(DE = case_when(
    padj >= 0.05 ~ "12951",
    log2FoldChange > 0 ~ "1279",
    log2FoldChange < 0 ~ "638"
  ) %>% factor(levels = c("1279","638","12951"))) %>%
  mutate(Significant=padj<0.05) %>%
  mutate(Significant = factor(Significant,levels = c(TRUE,FALSE))) %>%
  ggplot(aes(x=log2FoldChange,y=-log10(pvalue),color = DE)) +
  geom_hline(aes(yintercept = -log10(pForAlpha(resVenetoclax)),
             linetype = "padj = 0.05"),
             linewidth =2,
             color = "green") +
  geom_point(size = 2) +
  ylim(0,22.2) +
  scale_color_manual(values = c("1279" = "red","638" = "blue","12951" = "grey"),breaks = c("1279","638")) +
  scale_linetype_manual(name = "Cutoff",
                        values = c(2),
                        guide = guide_legend(override.aes = list(color = c("green")))) +
  guides(
    color = guide_legend(order = 1,
                         override.aes = list(size = 4)),
    linetype = guide_legend(order = 2)
  ) +
  theme_bw() +
  theme(text = element_text(face="bold",size = 32)) +
  ggtitle("VA/A D25")

```


```{r}
as.annot.data.frame(resVenetoclax,
                    subset = "All",
                    relabel = "Venetoclax Venetoclax vs None",
                    newlabel = "VA/A") %>%
  left_join(as.annot.data.frame(resAVvsA,
                                subset = "All",
                                relabel = "Group ARTposVenetoclax vs ART",
                                newlabel = "AV/A")) %>%
  left_join(as.annot.data.frame(resAVCvsA,
                                subset = "All",
                                relabel = "Group ARTposVenetoclaxposantiposCD8 vs ART",
                                newlabel = "AVC/A")) %>%
  left_join(as.annot.data.frame(resAVCvsAV,
                                subset = "All",
                                relabel = "Group ARTposVenetoclaxposantiposCD8 vs ARTposVenetoclax",
                                newlabel = "AVC/AV")) %>%
  openxlsx::write.xlsx(file = here("outputs/Merged_DESeq2_results_p24027_Tomas-20240930.xlsx"))
```


## GSEA


```{r GSEA_setup}
gmt.file <- list()

## See available genesets
# msigdbr_species()
# all_gene_sets = msigdbr(species = "Homo sapiens")
# head(all_gene_sets)

catlist <- c("H")
subcatlist <- c("CP:","GO:BP")

gmt.file <- sapply(filter(msigdbr_collections(),
              gs_cat %in% catlist |
                gs_subcat %in% (sapply(subcatlist,
                                       function(x){str_subset(msigdbr_collections() %>%
                                                                pull(gs_subcat),x)},
                                       USE.NAMES = TRUE) %>%
                                  unlist())) %>%
         split(1:7),
       function(x){msigdbr(species = "Homo sapiens",
                           category = x[1,]$gs_cat,
                           subcategory = x[1,]$gs_subcat) %>%
           dplyr::select(gene_symbol, gs_name)},simplify = FALSE) %>%
  bind_rows() %>%
  unstack()

# m_t2g_reactome <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME") %>% 
#   dplyr::select(gene_symbol, gs_name)
# m_t2g_biocarta <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:BIOCARTA") %>% 
#   dplyr::select(gene_symbol, gs_name)
# m_t2g_kegg <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>% 
#   dplyr::select(gene_symbol, gs_name)
# m_t2g_h <- msigdbr(species = "Homo sapiens", category = "H") %>% 
#   dplyr::select(gene_symbol, gs_name)
# gmt.file <- unstack(bind_rows(m_t2g_reactome,m_t2g_h,m_t2g_biocarta,m_t2g_kegg))

## Manually specify gmt.files
##gmt.file <- append(gmt.file, gmtPathways(file.path(a)))

bioplanet.pathways <- read_csv(here("config/pathway.csv"))
bioplanet.pathway.category <- read_csv(here("config/pathway-category.csv"))

bioplanet.gmt.file <- sapply(unique(bioplanet.pathways$PATHWAY_NAME),
                             function(pathway) {
                               filter(bioplanet.pathways, PATHWAY_NAME==pathway) %>%
                                 pull(GENE_SYMBOL)
                             },simplify = FALSE)


```


```{r}
resGSEAMsigdbVenetoclax <- fgsea(gmt.file,
                           as.data.frame(resVenetoclax) %>%
                             filter(!is.na(padj)) %>%
                             select(stat) %>%
                             rownames_to_column() %>%
                             pull(stat,name=rowname))
resGSEAMsigdbVenetoclax <- mutate(resGSEAMsigdbVenetoclax,
                                  source = str_extract(pathway,"[^_]*"),.before = pathway)
resGSEAMsigdbVenetoclax <- bind_cols(resGSEAMsigdbVenetoclax,
                                     tibble(LEn=sapply(resGSEAMsigdbVenetoclax$leadingEdge,length))) %>%
  relocate(c(1:8,10,9))

```

```{r}
arrange(resGSEAMsigdbVenetoclax,padj) %>%
  write.xlsx(here("outputs/Day25_VenetoclaxVsNone_GSEA_HC2CPGOBP_results_p24027_Tomas.xlsx"))
```

```{r fig.height=7}

sigHLMK <- filter(resGSEAMsigdbVenetoclax,source == "HALLMARK",pval < 0.05) %>%
  mutate(pathway = str_remove(pathway,pattern = "HALLMARK_")) %>%
  arrange(NES)

sigHLMK <- mutate(sigHLMK, pathway=factor(pathway,levels = sigHLMK$pathway))

sigHLMK %>%
  ggplot(aes(x=NES,y=pathway,color=pval,size=LEn)) +
  geom_point() +
  scale_color_gradient(low = "red",high = "blue") +
  ggtitle("Day25 Significant Hallmark") +
  theme_bw()
```

```{r}
heatmapWrapper <- function(geneList,
                                baseline_grouping = NULL,
                                baseline = NULL,
                                column_split = NULL,
                                slice_labels = NULL,
                                colors = c("blue", "white", "red"),
                                data = SummarizedExperiment::assays(analysis$dds)$rld,
                                column_labels = colnames(data),
                                slice_labels_rot = 90,
                                box_width = unit(3.5, "mm"),
                                box_height = unit(3.5, "mm"),
                                width_buffer = unit(5, "mm"),
                                height_buffer = unit(10, "mm"),
                                column_title = " ",
                                cluster_rows = FALSE,
                                cluster_columns = FALSE,
                                column_gap = unit(2, "mm"),
                                scale_min = -2,
                                scale_max = 2,
                                heatmap_legend_param = list(
                                  at = c(scale_min, 0, scale_max),
                                  labels = c(scale_min,0,scale_max),
                                  title = 'log2 fold\ndifference\nfrom\nmedian\nexpression'
                                ),
                                ...) {
  duds <- geneList[!geneList %in% rownames(data)]
  if (length(duds) > 0){
    geneList <- geneList[geneList %in% rownames(data)]
    if (length(geneList) == 0){
      stop('No data for requested genes')
    } else {
      message(paste0('Genes ', paste0(duds, collapse = ', '), ' not found in data'))
    }
  }
  hmap <- data[geneList, ]
  if (is.null(baseline_grouping) | is.null(baseline)) {
    message('Basline grouping or baseline level not specified, using all samples
            to generate median expression per gene')
    baseline <- matrixStats::rowMedians(SummarizedExperiment::assay(hmap))
  } else if (!baseline_grouping %in% colnames(colData(data))) {
    stop("Argument 'baseline_grouping' should be in colData(data)")
  } else if (!baseline %in% unique(colData(data)[[baseline_grouping]])) {
    stop("Argument 'baseline' should be a level of 'baseline_grouping' in colData(data)")
  } else{
    baseline <- matrixStats::rowMedians(SummarizedExperiment::assay(hmap[, as.character(hmap@colData[[baseline_grouping]]) %in% baseline]))
  }
  hmap <- SummarizedExperiment::assay(hmap) - baseline
  ComplexHeatmap::Heatmap(hmap,
    heatmap_legend_param = heatmap_legend_param,
    #border = "black",
    width = ncol(hmap) * box_width + width_buffer,
    height = nrow(hmap) * box_height + height_buffer,
    #rect_gp = grid::gpar(color = "black"),
    column_title = column_title,
    column_labels = column_labels,
    cluster_rows = cluster_rows,
    cluster_columns = cluster_columns,
    column_split = column_split,
    top_annotation = (if (!is.null(slice_labels)) {
      if (is.null(column_split)) {
        warning("Setting labels requires slices to also be set")
      }
      ComplexHeatmap::HeatmapAnnotation(foo = ComplexHeatmap::anno_block(
        gp = grid::gpar(col = NA),
        labels = slice_labels,
        labels_gp = grid::gpar(col = "black", fontsize = 10),
        labels_rot = slice_labels_rot, height = unit(2, "cm")
      ))
    } else {
      NULL
    }),
    column_gap = column_gap,
    col = circlize::colorRamp2(c(scale_min, 0, scale_max), colors),
    ...
  )
}

```


```{r fig.height=10}
heatmapWrapper(
  filter(resGSEAMsigdbVenetoclax,pathway == "HALLMARK_GLYCOLYSIS") %>%
    pull(leadingEdge) %>%
    unlist(),
  data = assays(analysis$dds)$rld,
  column_split = analysis$dds$Venetoclax,
  slice_labels = levels(analysis$dds$Venetoclax),
  slice_labels_rot = 0,
  column_title = "HALLMARK_GLYCOLYSIS",
  column_labels = analysis$dds$SubjectID,
  scale_min = -1,
  scale_max = 1
  )
```


```{r fig.height=10}
heatmapWrapper(
  filter(resGSEAMsigdbVenetoclax,pathway == "HALLMARK_APOPTOSIS") %>%
    pull(leadingEdge) %>%
    unlist(),
  data = assays(analysis$dds)$rld,
  column_split = analysis$dds$Venetoclax,
  slice_labels = levels(analysis$dds$Venetoclax),
  slice_labels_rot = 0,
  column_title = "HALLMARK_APOPTOSIS",
  column_labels = analysis$dds$SubjectID,
  scale_min = -1,
  scale_max = 1
  )
```


```{r}
resGSEAVenetoclax <- fgsea(bioplanet.gmt.file,
                           as.data.frame(resVenetoclax) %>%
                             filter(!is.na(padj)) %>%
                             select(stat) %>%
                             rownames_to_column() %>%
                             pull(stat,name=rowname))
```

```{r}
Tomas_setlistnames <- c("Toll receptor cascades",
                        "T cell receptor regulation of apoptosis",
                        "Interleukin-2 signaling pathway",
                        "Leptin influence on immune response",
                        "Cytokine-cytokine receptor interaction",
                        "Inflammasomes",
                        "NOD signaling pathway",
                        "Cross-presentation of particulate exogenous antigens (phagosomes)",
                        "Endogenous Toll-like receptor signaling",
                        "Lysosome",
                        "Lysosome vesicle biogenesis",
                        "Phagosome",
                        "Response to elevated platelet cytosolic calcium",
                        "Hemostasis pathway",
                        "Interleukin-5 regulation of apoptosis",
                        "Interleukin-4 regulation of apoptosis",
                        "TGF-beta regulation of extracellular matrix",
                        "Platelet activation, signaling and aggregation",
                        "Oncostatin M")

outlist <- list()
outlist$Selected_Day25_VvsNone <- filter(resGSEAVenetoclax,pathway %in% Tomas_setlistnames) %>% arrange(padj)
outlist$Full_Day25_VvsNone <- arrange(resGSEAVenetoclax,padj)

write.xlsx(outlist,here("outputs/Day25_VenetoclaxVsNone_GSEA_bioplanet_results_p24027_Tomas.xlsx"))
```


```{r}
extended.gmt.file <- c(set_names(bioplanet.gmt.file,str_c("bioplanet_",names(bioplanet.gmt.file))),gmt.file)

resGSEAList <- list(
  VAvsA.D25 = fgsea(extended.gmt.file,
                           as.data.frame(resVenetoclax) %>%
                             filter(!is.na(padj)) %>%
                             select(stat) %>%
                             rownames_to_column() %>%
                             pull(stat,name=rowname)) %>%
    mutate(source = str_extract(pathway,"[^_]*"),.before = pathway) %>%
    arrange(pval),
    AVvsA.D25 = fgsea(extended.gmt.file,
                           as.data.frame(resAVvsA) %>%
                             filter(!is.na(padj)) %>%
                             select(stat) %>%
                             rownames_to_column() %>%
                             pull(stat,name=rowname)) %>%
    mutate(source = str_extract(pathway,"[^_]*"),.before = pathway) %>%
    arrange(pval),
    AVCvsA.D25 = fgsea(extended.gmt.file,
                           as.data.frame(resAVCvsA) %>%
                             filter(!is.na(padj)) %>%
                             select(stat) %>%
                             rownames_to_column() %>%
                             pull(stat,name=rowname)) %>%
    mutate(source = str_extract(pathway,"[^_]*"),.before = pathway) %>%
    arrange(pval),
    AVCvsAV.D25 = fgsea(extended.gmt.file,
                           as.data.frame(resAVCvsAV) %>%
                             filter(!is.na(padj)) %>%
                             select(stat) %>%
                             rownames_to_column() %>%
                             pull(stat,name=rowname)) %>%
    mutate(source = str_extract(pathway,"[^_]*"),.before = pathway) %>%
    arrange(pval)
)


write.xlsx(resGSEAList,here("outputs/Day25_All_GSEA_MSigSB-bioplanet_results.xlsx"))

```



# Write tables

```{r}
writeCountTables(analysis,
                 normalized = FALSE,
                 generate_GSEA_app_tables = FALSE,
                 write_sample_table = TRUE)
```

# Save object

```{r}
if (!dir.exists(here('rds'))){dir.create(here('rds'))}
saveRDS(analysis, here('rds/analysis_post_QC_wSIV.Rds'))
```

# Render report

```{r}
rmarkdown::render(here::here('scripts/bulk_rnaseq_qc_template.format.Rmd'),
                  output_file = 'QC_report_wSIV.html',
                  output_dir = here::here('reports'),
                  params = list(
                    title = 'Bulk RNA QC Report')
                  )
```

